src/p11_child/p11_child_openssl.c

```diff
@@ -205,6 +206,72 @@ static const EVP_MD *get_dgst(CK_MECHANISM_TYPE ocsp_dgst)
     return dgst;
 }
 
+static char *get_issuer_subject_str(TALLOC_CTX *mem_ctx, X509 *cert)
+{
+    X509_NAME *issuer_name;
+    X509_NAME *subject_name;
+    char *tmp_str = NULL;
+    BIO *bio_mem = NULL;
+    int ret;
+    char *str = NULL;
+
+    issuer_name = X509_get_issuer_name(cert);
+    subject_name = X509_get_subject_name(cert);
+
+    if (issuer_name == NULL || subject_name == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing issuer or subject.\n");
+        return NULL;
+    }
+
+    bio_mem = BIO_new(BIO_s_mem());
+    if (bio_mem == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "BIO_new failed.\n");
+        return NULL;
+    }
+
+    ret = BIO_printf(bio_mem, "Issuer: [");
+    if (ret == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "BIO_printf failed.\n");
+        goto done;
+    }
+
+    ret = X509_NAME_print_ex(bio_mem, issuer_name, 0, XN_FLAG_ONELINE);
+    if (ret == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "X509_NAME_print_ex failed.\n");
+        goto done;
+    }
+
+    ret = BIO_printf(bio_mem, "] Subject: [");
+    if (ret == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "BIO_printf failed.\n");
+        goto done;
+    }
+
+    ret = X509_NAME_print_ex(bio_mem, subject_name, 0, XN_FLAG_ONELINE);
+    if (ret == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "X509_NAME_print_ex failed.\n");
+        goto done;
+    }
+
+    ret = BIO_printf(bio_mem, "]");
+    if (ret == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "BIO_printf failed.\n");
+        goto done;
+    }
+
+    if (BIO_get_mem_data(bio_mem, &tmp_str) == 0) {
```

Comment:
According to the [man page](https://www.openssl.org/docs/man1.1.1/man3/BIO_get_mem_data.html):
```
BIO_get_mem_data() returns the total number of bytes available on success, 
0 if b is NULL, or a negative value in case of other errors.
```
So I think fail-check should be `<=0`

Reply:
For the record: I was wondering what assures that BIO buffer contains null-terminated string in this case. Man pages are quite vague here just saying "similar to printf"
But it seems implementation indeed [puts trailing null](https://github.com/openssl/openssl/blob/master/crypto/bio/b_print.c#L357).

Reply:
That's a good point. I used the length returned by `BIO_get_mem_data()` to make sure that only the number of expected characters are copied and fixed the erro check, too.

Reply:
Thank you.

While `_dopr()` puts trailing 0 in a temp buffer, it also returns `*retlen = currlen - 1` and later `BIO_vprintf()` does `BIO_write(bio, ..., (int)retlen)`

So seems like this trailing 0 doesn't get into final buffer but I am not sure. It is better to be on a safe side.